In [1]:
import polars as pl

Link to dataset: [https://www.kaggle.com/datasets/undefinenull/million-song-dataset-spotify-lastfm](https://www.kaggle.com/datasets/undefinenull/million-song-dataset-spotify-lastfm)

In [2]:
tracks_lf = pl.scan_csv('data/Music Info.csv')

In [3]:
tracks_lf.fetch(5)

shape: (5, 21)
┌────────────┬────────────┬───────────┬────────────┬───┬──────────┬─────────┬─────────┬────────────┐
│ track_id   ┆ name       ┆ artist    ┆ spotify_pr ┆ … ┆ liveness ┆ valence ┆ tempo   ┆ time_signa │
│ ---        ┆ ---        ┆ ---       ┆ eview_url  ┆   ┆ ---      ┆ ---     ┆ ---     ┆ ture       │
│ str        ┆ str        ┆ str       ┆ ---        ┆   ┆ f64      ┆ f64     ┆ f64     ┆ ---        │
│            ┆            ┆           ┆ str        ┆   ┆          ┆         ┆         ┆ i64        │
╞════════════╪════════════╪═══════════╪════════════╪═══╪══════════╪═════════╪═════════╪════════════╡
│ TRIOREW128 ┆ Mr.        ┆ The       ┆ https://p. ┆ … ┆ 0.0971   ┆ 0.24    ┆ 148.114 ┆ 4          │
│ F424EAF0   ┆ Brightside ┆ Killers   ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 4d…        ┆   ┆          ┆         ┆         ┆            │
│ TRRIVDJ128 ┆ Wonderwall ┆ Oasis     ┆ https://p. ┆ … ┆ 0.207    ┆ 0.651   ┆ 174.426 ┆ 4          │
│ F429B0E8   ┆            ┆           ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ d0…        ┆   ┆          ┆         ┆         ┆            │
│ TROUVHL128 ┆ Come as    ┆ Nirvana   ┆ https://p. ┆ … ┆ 0.0878   ┆ 0.543   ┆ 120.012 ┆ 4          │
│ F426C441   ┆ You Are    ┆           ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ a1…        ┆   ┆          ┆         ┆         ┆            │
│ TRUEIND128 ┆ Take Me    ┆ Franz     ┆ https://p. ┆ … ┆ 0.133    ┆ 0.49    ┆ 104.56  ┆ 4          │
│ F93038C4   ┆ Out        ┆ Ferdinand ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 39…        ┆   ┆          ┆         ┆         ┆            │
│ TRLNZBD128 ┆ Creep      ┆ Radiohead ┆ https://p. ┆ … ┆ 0.129    ┆ 0.104   ┆ 91.841  ┆ 4          │
│ F935E4D8   ┆            ┆           ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ e7…        ┆   ┆          ┆         ┆         ┆            │
└────────────┴────────────┴───────────┴────────────┴───┴──────────┴─────────┴─────────┴────────────┘

In [4]:
cols = [
    (100 - pl.col(col).is_null().sum() * 100 / pl.col(col).count())
        .round(2)
        .alias('{}_full %'.format(col))
    for col in tracks_lf.columns
]

tracks_lf.select(cols).collect().transpose(include_header=True, column_names=['full %'])

shape: (21, 2)
┌────────────────────────────┬────────┐
│ column                     ┆ full % │
│ ---                        ┆ ---    │
│ str                        ┆ f64    │
╞════════════════════════════╪════════╡
│ track_id_full %            ┆ 100.0  │
│ name_full %                ┆ 100.0  │
│ artist_full %              ┆ 100.0  │
│ spotify_preview_url_full % ┆ 100.0  │
│ …                          ┆ …      │
│ liveness_full %            ┆ 100.0  │
│ valence_full %             ┆ 100.0  │
│ tempo_full %               ┆ 100.0  │
│ time_signature_full %      ┆ 100.0  │
└────────────────────────────┴────────┘

In [5]:
tracks_lf.select(pl.all()).filter(pl.col('genre').is_null()).fetch(10)

shape: (7, 21)
┌────────────┬────────────┬───────────┬────────────┬───┬──────────┬─────────┬─────────┬────────────┐
│ track_id   ┆ name       ┆ artist    ┆ spotify_pr ┆ … ┆ liveness ┆ valence ┆ tempo   ┆ time_signa │
│ ---        ┆ ---        ┆ ---       ┆ eview_url  ┆   ┆ ---      ┆ ---     ┆ ---     ┆ ture       │
│ str        ┆ str        ┆ str       ┆ ---        ┆   ┆ f64      ┆ f64     ┆ f64     ┆ ---        │
│            ┆            ┆           ┆ str        ┆   ┆          ┆         ┆         ┆ i64        │
╞════════════╪════════════╪═══════════╪════════════╪═══╪══════════╪═════════╪═════════╪════════════╡
│ TRIOREW128 ┆ Mr.        ┆ The       ┆ https://p. ┆ … ┆ 0.0971   ┆ 0.24    ┆ 148.114 ┆ 4          │
│ F424EAF0   ┆ Brightside ┆ Killers   ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 4d…        ┆   ┆          ┆         ┆         ┆            │
│ TRRIVDJ128 ┆ Wonderwall ┆ Oasis     ┆ https://p. ┆ … ┆ 0.207    ┆ 0.651   ┆ 174.426 ┆ 4          │
│ F429B0E8   ┆            ┆           ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ d0…        ┆   ┆          ┆         ┆         ┆            │
│ TRUEIND128 ┆ Take Me    ┆ Franz     ┆ https://p. ┆ … ┆ 0.133    ┆ 0.49    ┆ 104.56  ┆ 4          │
│ F93038C4   ┆ Out        ┆ Ferdinand ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 39…        ┆   ┆          ┆         ┆         ┆            │
│ TRUMISQ128 ┆ Somebody   ┆ The       ┆ https://p. ┆ … ┆ 0.0641   ┆ 0.704   ┆ 138.03  ┆ 4          │
│ F9340BEE   ┆ Told Me    ┆ Killers   ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 0d…        ┆   ┆          ┆         ┆         ┆            │
│ TRVCCWR128 ┆ Viva la    ┆ Coldplay  ┆ https://p. ┆ … ┆ 0.0634   ┆ 0.52    ┆ 137.973 ┆ 4          │
│ F9304A30   ┆ Vida       ┆           ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ ab…        ┆   ┆          ┆         ┆         ┆            │
│ TRXOGZT128 ┆ Karma      ┆ Radiohead ┆ https://p. ┆ … ┆ 0.172    ┆ 0.317   ┆ 74.807  ┆ 4          │
│ F424AD74   ┆ Police     ┆           ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 5a…        ┆   ┆          ┆         ┆         ┆            │
│ TRUJIIV129 ┆ Clocks     ┆ Coldplay  ┆ https://p. ┆ … ┆ 0.183    ┆ 0.255   ┆ 130.97  ┆ 4          │
│ 03CA8848   ┆            ┆           ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 24…        ┆   ┆          ┆         ┆         ┆            │
└────────────┴────────────┴───────────┴────────────┴───┴──────────┴─────────┴─────────┴────────────┘

In [6]:
tracks_lf.select(pl.col('genre').value_counts()).collect()

shape: (16, 1)
┌─────────────────────┐
│ genre               │
│ ---                 │
│ struct[2]           │
╞═════════════════════╡
│ {"RnB",696}         │
│ {"Rap",821}         │
│ {"Reggae",691}      │
│ {"Folk",355}        │
│ …                   │
│ {"New Age",237}     │
│ {"Punk",383}        │
│ {"Electronic",3710} │
│ {"Latin",100}       │
└─────────────────────┘

In [7]:
tracks_lf = tracks_lf.select(pl.all()).with_columns(pl.col('tags').str.split(', ').alias('tags'))

In [8]:
def tag_equals_to_genre() -> pl.expr.expr.Expr:
    return (
        pl.col('tags').arr.get(0).str.to_uppercase()
        ==
        pl.col('genre').str.to_uppercase()
    )

In [9]:
tracks_lf.drop_nulls().select(pl.all()).filter(tag_equals_to_genre().is_not()).collect()

shape: (12_678, 21)
┌────────────┬────────────┬───────────┬────────────┬───┬──────────┬─────────┬─────────┬────────────┐
│ track_id   ┆ name       ┆ artist    ┆ spotify_pr ┆ … ┆ liveness ┆ valence ┆ tempo   ┆ time_signa │
│ ---        ┆ ---        ┆ ---       ┆ eview_url  ┆   ┆ ---      ┆ ---     ┆ ---     ┆ ture       │
│ str        ┆ str        ┆ str       ┆ ---        ┆   ┆ f64      ┆ f64     ┆ f64     ┆ ---        │
│            ┆            ┆           ┆ str        ┆   ┆          ┆         ┆         ┆ i64        │
╞════════════╪════════════╪═══════════╪════════════╪═══╪══════════╪═════════╪═════════╪════════════╡
│ TROUVHL128 ┆ Come as    ┆ Nirvana   ┆ https://p. ┆ … ┆ 0.0878   ┆ 0.543   ┆ 120.012 ┆ 4          │
│ F426C441   ┆ You Are    ┆           ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ a1…        ┆   ┆          ┆         ┆         ┆            │
│ TRLNZBD128 ┆ Creep      ┆ Radiohead ┆ https://p. ┆ … ┆ 0.129    ┆ 0.104   ┆ 91.841  ┆ 4          │
│ F935E4D8   ┆            ┆           ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ e7…        ┆   ┆          ┆         ┆         ┆            │
│ TRIODZU128 ┆ Under the  ┆ Red Hot   ┆ https://p. ┆ … ┆ 0.136    ┆ 0.513   ┆ 84.275  ┆ 4          │
│ E078F3E2   ┆ Bridge     ┆ Chili     ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆ Peppers   ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 90…        ┆   ┆          ┆         ┆         ┆            │
│ TRYEZGE128 ┆ Feel Good  ┆ Gorillaz  ┆ https://p. ┆ … ┆ 0.826    ┆ 0.832   ┆ 138.556 ┆ 4          │
│ F92D04C8   ┆ Inc.       ┆           ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 73…        ┆   ┆          ┆         ┆         ┆            │
│ …          ┆ …          ┆ …         ┆ …          ┆ … ┆ …        ┆ …       ┆ …       ┆ …          │
│ TRUVVQK128 ┆ PLAY THE   ┆ ストレイ  ┆ https://p. ┆ … ┆ 0.346    ┆ 0.362   ┆ 177.915 ┆ 3          │
│ F4285D05   ┆ STAR       ┆ テナー    ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆ GUITAR     ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ b2…        ┆   ┆          ┆         ┆         ┆            │
│ TRTBIEH128 ┆ Let It All ┆ Boom Boom ┆ https://p. ┆ … ┆ 0.164    ┆ 0.118   ┆ 103.008 ┆ 4          │
│ F427C13C   ┆ Come Down  ┆ Satellite ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆ s         ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 5d…        ┆   ┆          ┆         ┆         ┆            │
│ TRPIGDW129 ┆ Slip of    ┆ Fact      ┆ https://p. ┆ … ┆ 0.379    ┆ 0.224   ┆ 100.08  ┆ 4          │
│ 03CDEB2D   ┆ the Lip    ┆           ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ cf…        ┆   ┆          ┆         ┆         ┆            │
│ TRIPFKO128 ┆ Hide &     ┆ 9mm Parab ┆ https://p. ┆ … ┆ 0.203    ┆ 0.138   ┆ 136.475 ┆ 4          │
│ F42383FE   ┆ Seek       ┆ ellum     ┆ scdn.co/mp ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆ Bullet    ┆ 3-preview/ ┆   ┆          ┆         ┆         ┆            │
│            ┆            ┆           ┆ 91…        ┆   ┆          ┆         ┆         ┆            │
└────────────┴────────────┴───────────┴─────────

Something seems to be off with genre column - its content does not match many of the songs.
It would be instead more productive to treat subset of tags as genres, since they contain far more specific and accurate information. Multiple genres instead of one will also suport describing songs, that were more experimental and do not adhere to any single label.

In [10]:
def to_snake(text: str) -> str:
    return text.replace(' ', '_').lower()

def genre_is_tag() -> pl.expr.expr.Expr:
    return pl.col('genre').apply(to_snake).is_in(pl.col('tags'))

In [11]:
tracks_lf.drop_nulls().select(
    genre_is_tag().sum()
    /
    pl.col('genre').count()
).collect()

shape: (1, 1)
┌──────────┐
│ genre    │
│ ---      │
│ f64      │
╞══════════╡
│ 0.474265 │
└──────────┘

In [12]:
tracks_lf = pl.concat([
    tracks_lf.select(
        pl.all()
    ).filter(
        pl.col('genre').is_not_null()
    ).with_columns(
        pl.col('tags')
            .arr.concat(pl.col('genre').apply(to_snake, skip_nulls=True))
            .arr.unique()
            .alias('tags')
    ),
    tracks_lf.select(
        pl.all()
    ).filter(
        pl.col('genre').is_null()
    )
])

In [13]:
tracks_lf.drop_nulls().select(
    genre_is_tag().sum()
    /
    pl.col('genre').count()
).collect()

shape: (1, 1)
┌───────┐
│ genre │
│ ---   │
│ f64   │
╞═══════╡
│ 1.0   │
└───────┘

In [14]:
tracks_genreless_lf = tracks_lf.select(pl.all().exclude('genre'))

In [15]:
tag_list = tracks_lf.select(pl.col('tags').drop_nulls().explode().drop_nulls()).unique().collect().to_numpy().flatten().tolist()
print(tag_list)
with open('tmp/tag-list.txt', 'w+') as f:
    f.write('\n'.join(tag_list))

['electro', 'latin', 'instrumental', 'nu_metal', 'indie_pop', '70s', 'acoustic', 'avant_garde', 'rock', 'rnb', 'blues', 'german', '90s', 'classic_rock', 'techno', 'new_age', 'psychedelic', 'folk', 'hard_rock', 'dance', 'american', 'beautiful', 'experimental', 'grunge', 'progressive_rock', 'oldies', 'thrash_metal', 'female_vocalists', 'post_punk', 'french', 'melodic_death_metal', 'reggae', 'punk_rock', 'progressive_metal', 'symphonic_metal', '60s', 'post_hardcore', 'mellow', 'lounge', 'ambient', 'post_rock', 'indie_rock', 'swedish', 'gothic', 'emo', 'russian', 'hip_hop', 'heavy_metal', 'classical', 'british', 'house', 'blues_rock', '80s', 'indie', 'hardcore', 'idm', 'noise', 'pop_rock', 'chillout', 'polish', 'world', 'psychedelic_rock', 'trip_hop', 'trance', 'metal', 'dark_ambient', 'punk', 'grindcore', 'electronic', 'chill', 'gothic_metal', 'male_vocalists', 'soul', 'pop', 'death_metal', 'downtempo', 'guitar', 'cover', 'synthpop', 'metalcore', 'jazz', 'rap', 'screamo', 'funk', 'love', 

With expert knowledge from outside sources, such as rateyourmusic.com, I was able to indentificate non-genre tags and remove them. It shortened the list from 102 to 77.

In [16]:
with open('tmp/tag-list-filtered.txt') as f:
    genre_list = f.read().split('\n')
print(genre_list)

['nu_metal', 'trance', 'trip_hop', 'instrumental', 'gothic', 'rock', 'blues_rock', 'electro', 'french', 'folk', 'j_pop', 'pop', 'dark_ambient', 'post_rock', 'guitar', 'progressive_metal', 'house', 'american', 'grunge', 'punk', 'lounge', 'psychedelic', 'experimental', 'hip_hop', 'melodic_death_metal', 'rnb', 'classic_rock', 'screamo', 'punk_rock', 'emo', 'thrash_metal', 'death_metal', 'blues', 'symphonic_metal', 'grindcore', 'classical', 'doom_metal', 'gothic_metal', 'post_hardcore', 'indie', 'rap', 'post_punk', 'hard_rock', 'new_wave', 'new_age', 'idm', 'power_metal', 'avant_garde', 'britpop', 'hardcore', 'techno', 'jazz', 'electronic', 'industrial', 'reggae', 'indie_pop', 'drum_and_bass', 'ambient', 'ska', 'country', 'indie_rock', 'soul', 'dance', 'metalcore', 'metal', 'black_metal', 'downtempo', 'funk', 'synthpop', 'psychedelic_rock', 'noise', 'alternative', 'heavy_metal', 'progressive_rock', 'pop_rock', 'alternative_rock', 'soundtrack']


In [17]:
def tag_not_in_genre_list():
    return pl.col('tags').arr.eval(
        pl.element().is_in(genre_list).any().is_not()
    ).arr.get(0)

tracks_incorrect_df = tracks_genreless_lf.select(
    pl.col('name'),
    pl.col('artist'),
    pl.col('tags')
).filter(
    tag_not_in_genre_list()
).collect()
tracks_incorrect_df

shape: (602, 3)
┌──────────────────────┬───────────────────────┬───────────────────────────────┐
│ name                 ┆ artist                ┆ tags                          │
│ ---                  ┆ ---                   ┆ ---                           │
│ str                  ┆ str                   ┆ list[str]                     │
╞══════════════════════╪═══════════════════════╪═══════════════════════════════╡
│ Dune                 ┆ L'Arc~en~Ciel         ┆ ["japanese", "world"]         │
│ El Triste            ┆ José José             ┆ ["oldies", "love", "latin"]   │
│ Te Lo Pido Por Favor ┆ Jaguares              ┆ ["cover", "latin"]            │
│ Petit Pays           ┆ Cesária Évora         ┆ ["female_vocalists", "world"] │
│ …                    ┆ …                     ┆ …                             │
│ Battle March         ┆ 9mm Parabellum Bullet ┆ ["japanese"]                  │
│ Kakusei Heroism      ┆ アンティック-珈琲店-  ┆ ["japanese"]                  │
│ Lights Of Vein       ┆ Fact                  ┆ ["japanese"]                  │
│ BERSERKER TUNE       ┆ ストレイテナー        ┆ ["japanese"]                  │
└──────────────────────┴───────────────────────┴───────────────────────────────┘

In [18]:
tracks_incorrect_df.select(pl.first().count().alias('count'))

shape: (1, 1)
┌───────┐
│ count │
│ ---   │
│ u32   │
╞═══════╡
│ 602   │
└───────┘

For now, I will leave out the 602 genre-less tracks.

In [19]:
tracks_filtered_lf = tracks_genreless_lf.select([
    pl.all().exclude('tags'),
    pl.col('tags').arr.join(';').alias('genres')
]).filter(
    tag_not_in_genre_list().is_not()
)

In [20]:
listens_lf = pl.scan_csv('data/User Listening History.csv')

In [21]:
listens_count_lf = listens_lf.select([
    pl.col('track_id'),
    pl.col('playcount')
]).groupby('track_id').sum()

In [22]:
tracks_with_listens_df = tracks_filtered_lf.join(listens_count_lf, on='track_id').collect()
tracks_with_listens_df

shape: (29_477, 21)
┌───────────┬───────────┬───────────┬────────────┬───┬─────────┬────────────┬──────────┬───────────┐
│ track_id  ┆ name      ┆ artist    ┆ spotify_pr ┆ … ┆ tempo   ┆ time_signa ┆ genres   ┆ playcount │
│ ---       ┆ ---       ┆ ---       ┆ eview_url  ┆   ┆ ---     ┆ ture       ┆ ---      ┆ ---       │
│ str       ┆ str       ┆ str       ┆ ---        ┆   ┆ f64     ┆ ---        ┆ str      ┆ i64       │
│           ┆           ┆           ┆ str        ┆   ┆         ┆ i64        ┆          ┆           │
╞═══════════╪═══════════╪═══════════╪════════════╪═══╪═════════╪════════════╪══════════╪═══════════╡
│ TRIODZU12 ┆ Under the ┆ Red Hot   ┆ https://p. ┆ … ┆ 84.275  ┆ 4          ┆ rock;alt ┆ 990       │
│ 8E078F3E2 ┆ Bridge    ┆ Chili     ┆ scdn.co/mp ┆   ┆         ┆            ┆ ernative ┆           │
│           ┆           ┆ Peppers   ┆ 3-preview/ ┆   ┆         ┆            ┆ ;alterna ┆           │
│           ┆           ┆           ┆ 90…        ┆   ┆         ┆            ┆ tive_roc ┆           │
│           ┆           ┆           ┆            ┆   ┆         ┆            ┆ …        ┆           │
│ TRYEZGE12 ┆ Feel Good ┆ Gorillaz  ┆ https://p. ┆ … ┆ 138.556 ┆ 4          ┆ alternat ┆ 98        │
│ 8F92D04C8 ┆ Inc.      ┆           ┆ scdn.co/mp ┆   ┆         ┆            ┆ ive_rock ┆           │
│           ┆           ┆           ┆ 3-preview/ ┆   ┆         ┆            ┆ ;british ┆           │
│           ┆           ┆           ┆ 73…        ┆   ┆         ┆            ┆ ;trip_ho ┆           │
│           ┆           ┆           ┆            ┆   ┆         ┆            ┆ …        ┆           │
│ TRIQWYH12 ┆ Paranoid  ┆ Radiohead ┆ https://p. ┆ … ┆ 81.849  ┆ 4          ┆ britpop; ┆ 1075      │
│ 8F1486062 ┆ Android   ┆           ┆ scdn.co/mp ┆   ┆         ┆            ┆ progress ┆           │
│           ┆           ┆           ┆ 3-preview/ ┆   ┆         ┆            ┆ ive_rock ┆           │
│           ┆           ┆           ┆ d6…        ┆   ┆         ┆            ┆ ;90s;ele ┆           │
│           ┆           ┆           ┆            ┆   ┆         ┆            ┆ …        ┆           │
│ TRUOEWJ12 ┆ Stairway  ┆ Led       ┆ https://p. ┆ … ┆ 80.36   ┆ 4          ┆ 70s;punk ┆ 566       │
│ 8F9345A85 ┆ to Heaven ┆ Zeppelin  ┆ scdn.co/mp ┆   ┆         ┆            ┆ ;hard_ro ┆           │
│           ┆           ┆           ┆ 3-preview/ ┆   ┆         ┆            ┆ ck;progr ┆           │
│           ┆           ┆           ┆ 68…        ┆   ┆         ┆            ┆ essive_r ┆           │
│           ┆           ┆           ┆            ┆   ┆         ┆            ┆ …        ┆           │
│ …         ┆ …         ┆ …         ┆ …          ┆ … ┆ …       ┆ …          ┆ …        ┆ …         │
│ TRXJFMF12 ┆ Living    ┆ 9mm Parab ┆ https://p. ┆ … ┆ 95.001  ┆ 4          ┆ rock     ┆ 11        │
│ 8F92E455D ┆ Dying     ┆ ellum     ┆ scdn.co/mp ┆   ┆         ┆            ┆          ┆           │
│           ┆ Message   ┆ Bullet    ┆ 3-preview/ ┆   ┆         ┆            ┆          ┆           │
│           ┆           ┆           ┆ 56…        ┆   ┆         ┆            ┆          ┆           │
│ TRMEHFD12 ┆ The       ┆ 9mm Parab ┆ https://p. ┆ … ┆ 192.071 ┆ 4          ┆ rock;ins ┆ 42        │
│ 8F92E4557 ┆ Revenge   ┆ ellum     ┆ scdn.co/mp ┆   ┆         ┆            ┆ trumenta ┆           │
│           ┆ of Surf   ┆ Bullet    ┆ 3-preview/ ┆   ┆         ┆            ┆ l;japane ┆           │
│           ┆ Queen     ┆           ┆ 0e…        ┆   ┆         ┆            ┆ se       ┆           │
│ TRXWSIN12 ┆ Infinite  ┆ Maximilia ┆ https://p. ┆ … ┆ 99.679  ┆ 4          ┆ electron ┆ 36        │
│ 8F9339A11 ┆ Love Song ┆ n Hecker  ┆ scdn.co/mp ┆   ┆         ┆            ┆ ic;indie ┆           │
│           ┆           ┆           ┆ 3-preview/ ┆   ┆         ┆            ┆ _pop;lov ┆           │
│           ┆           ┆           ┆ 8b…        ┆   ┆         ┆            ┆ e        ┆           │
│ TRHQCSH12 ┆ Colors Of ┆ ACIDMAN   ┆ htt

In [23]:
tracks_with_listens_df.write_csv('data/music-info-clean.csv')